### PR TITLE
Add VLOOKUP formula

### PR DESCRIPTION
**Pull request by marthacryan**
Adds support for the VLOOKUP formula within the same sheet. Cross-sheet VLOOKUP to come! 

**Todo and questions**:
- [x] Add more tests
- [ ] Update the behavior for other lookup calls. **Should they stay the same (throw an error to specifically use the merge functionality) or should they point to the VLOOKUP formula instead?**
- [x] What is the behavior if the *lookup value* is a primitive and the *column indices* are a list? 
- [x] Cast to the correct types?
- [x] make sure to only take the first occurence of the lookup value

### CODE REVIEW

File: mitosheet/mitosheet/tests/public/v3/sheet_functions/misc_functions/test_vlookup.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        ],
+        pd.Series(['d', None, 'e'])
+    ),
+]
+
+@pytest.mark.parametrize("_argv, expected", TYPE_VALID_TESTS)
+def test_vlookup_direct(_argv, expected):
+    result = VLOOKUP(*_argv)
+    print(f'expected: {expected}')
+    print(f'result: {result}')
```
Comment (naterush) [inline mitosheet/mitosheet/tests/public/v3/sheet_functions/misc_functions/test_vlookup.py:94]: Cleanup

File: mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py [hunk 2]
```diff
@@ -149,11 +150,59 @@ def GETNEXTVALUE(series: pd.Series, condition: BoolRestrictedInputType) -> pd.Se
...
+    where_deduplicated = where.drop_duplicates(subset=where.iloc[:,0].name)
+    merged = pd.merge(value, where_deduplicated, left_on='lookup_value', right_on=where_deduplicated.iloc[:,0], how='left')
+    if isinstance(index, int):
+        return merged.iloc[:, index]
+    else:
+        def get_nth_indexed_column(row):
+            try:
+                return row[index[row.name]-1]
+            except Exception as e:
+                print(f"error: {e}")
```
Comment (naterush) [inline mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py:197]: Remove print

File: mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py [hunk 3]
```diff
@@ -149,11 +150,59 @@ def GETNEXTVALUE(series: pd.Series, condition: BoolRestrictedInputType) -> pd.Se
...
+    
+    # If the lookup value and index are both a primitive, we don't need to merge. 
+    if not isinstance(lookup_value, pd.Series) and isinstance(index, int):
+        return where.loc[where.iloc[:,0] == lookup_value].iloc[0, index-1]
+    
+    value = get_series_from_primitive_or_series(lookup_value, where.index)
+    value.name = 'lookup_value'
+    
+    # Then we want to do a merge on the column we're looking up from, and the df we're looking up in.
+    where_deduplicated = where.drop_duplicates(subset=where.iloc[:,0].name)
```
Comment (naterush) [inline mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py:188]: might be helpful to save `where.iloc[:,0]` as `where_first_column` for readability - took me a bit to parse.

File: mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py [hunk 4]
```diff
@@ -149,11 +150,59 @@ def GETNEXTVALUE(series: pd.Series, condition: BoolRestrictedInputType) -> pd.Se
...
+    merged = pd.merge(value, where_deduplicated, left_on='lookup_value', right_on=where_deduplicated.iloc[:,0], how='left')
+    if isinstance(index, int):
+        return merged.iloc[:, index]
+    else:
+        def get_nth_indexed_column(row):
```
Comment (naterush) [inline mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py:193]: This function name is a tad confusing. Can we think of something more clear. Maybe like `get_value_at_index_in_row` or something?

File: mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py [hunk 5]
```diff
@@ -149,11 +150,59 @@ def GETNEXTVALUE(series: pd.Series, condition: BoolRestrictedInputType) -> pd.Se
 
     return GETPREVIOUSVALUE(reversed_series, reversed_condition)[::-1]
 
+@cast_values_in_arg_to_type('index', 'int')
+def VLOOKUP(lookup_value: AnyPrimitiveOrSeriesInputType, where: pd.DataFrame, index: IntRestrictedInputType) -> pd.Series:
+    """
+    {
+        "function": "VLOOKUP",
+        "description": "Returns the next value from series that meets the condition.",
```
Comment (naterush) [inline mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py:158]: This needs updating, I think.

File: mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py [hunk 6]
```diff
@@ -149,11 +150,59 @@ def GETNEXTVALUE(series: pd.Series, condition: BoolRestrictedInputType) -> pd.Se
...
+        "search_terms": ["vlookup", "merge", "join", "search", "lookup"],
+        "examples": [
+            "VLOOKUP(Names0, Sample_Names:Sample_Ages, 1, False)",
```
Comment (naterush) [inline mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py:161]: Maybe we could pick more informative sample values. Like `ids` as the columns or something. And the where is `ids:age` or something. Sample is kinda a filler word.

**Review by naterush — COMMENTED** [review]
A few more tests, and a bit of cleanup, and then LGTM! Good work on this one :)

File: mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py [hunk 7]
```diff
@@ -149,11 +150,59 @@ def GETNEXTVALUE(series: pd.Series, condition: BoolRestrictedInputType) -> pd.Se
...
+    merged = pd.merge(value, where_deduplicated, left_on='lookup_value', right_on=where_deduplicated.iloc[:,0], how='left')
+    if isinstance(index, int):
+        return merged.iloc[:, index]
+    else:
+        def get_nth_indexed_column(row):
+            try:
+                return row[index[row.name]-1]
+            except Exception as e:
+                print(f"error: {e}")
+                return None
```
[the remark this answers is not in the document]
Comment (marthacryan) [reply mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py:218]: The index is user inputted, so we might be getting a column that's out of range from the user. When they do this, I think it makes sense to just return None.

[the remark this answers is not in the document]
Comment (marthacryan) [reply mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py:218]: I added a comment to explain!

File: mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py (hunk 6 above)
Comment (marthacryan) [reply mitosheet/mitosheet/public/v3/sheet_functions/misc_functions.py]: How does that updated version look?

**Pull request merged**